pkg/controller/pravegacluster/pravegacluster_finalizer.go

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright (c) 2018 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package pravegacluster
+
+import (
+	"context"
+	"fmt"
+	pravegav1alpha1 "github.com/pravega/pravega-operator/pkg/apis/pravega/v1alpha1"
+	"github.com/pravega/pravega-operator/pkg/util"
+	"k8s.io/apimachinery/pkg/labels"
+	"os"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/remotecommand"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ZKMETAFINALIZER = "zk.meta.finalizer.pravega.io"
+	ZKUSERFINALIZER = "zk.user.finalizer.pravega.io"
+)
+
+func (r *ReconcilePravegaCluster) cleanUpMetaInZk(p *pravegav1alpha1.PravegaCluster) (err error) {
+	// Find a zookeeper node by using the selector in zookeeper service
+	zkService := &corev1.Service{}
+	buf := strings.Split(p.Spec.ZookeeperUri, ":")
+	zkServiceName := buf[0]
+	err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: p.Namespace, Name: zkServiceName}, zkService)
+	if err != nil {
+		return fmt.Errorf("failed to get zookeeper service (%s): %v", zkServiceName, err)
+	}
+
+	zkNodeLabels := util.LabelsForZookeeperNode(zkService.Spec.Selector)
+	zkNodeSelector := labels.SelectorFromSet(zkNodeLabels)
+	zkNodeListOps := &client.ListOptions{Namespace: p.Namespace, LabelSelector: zkNodeSelector}
+
+	zkNodeList := &corev1.PodList{}
+	err = r.client.List(context.TODO(), zkNodeListOps, zkNodeList)
+	if err != nil {
+		return fmt.Errorf("failed to list zookeeper pods: %v", err)
+	}
+
+	if len(zkNodeList.Items) == 0 {
+		return fmt.Errorf("failed to find zookeeper node")
+	}
+
+	// Choose one of the zookeeper nodes
+	zkNode := zkNodeList.Items[0]
+	kubeconfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		&clientcmd.ConfigOverrides{},
+	)
+	restconfig, err := kubeconfig.ClientConfig()
+	if err != nil {
+		return fmt.Errorf("failed to get REST config from kube config: %v", err)
+	}
+
+	coreclient, err := corev1client.NewForConfig(restconfig)
+	if err != nil {
+		return fmt.Errorf("failed to create client from REST config: %v", err)
+	}
+
+	req := coreclient.RESTClient().
```

Comment:
Let's explore the use of a Zookeeper Go library and use the Service address instead of `exec` into one of the nodes.

Reply:
After researching, I cannot find a zk client that can delete znode recursively. The only way to delete all is to use zkCli.sh, which needs to log in one of the nodes. If we want to use Go library, we will need to write a BFS to delete znode one by one and I suppose that will cause a lot of network traffics. What do you think?

Reply:
I am using [go-zookeeper](https://github.com/samuel/go-zookeeper) now